modules/ml/src/main/java/org/apache/ignite/ml/tree/randomforest/RandomForestTrainer.java

```diff
@@ -17,45 +17,815 @@
+            return new NormalDistributionStats(
+                Math.min(this.min, stats.min),
+                Math.max(this.max, stats.max),
+                this.sumOfSquares + stats.sumOfSquares,
+                this.sumOfValues + stats.sumOfValues,
+                this.n + stats.n
+            );
+        }
+
+        /**
+         * Mean value.
+         */
+        public double mean() {
+            return sumOfValues / n;
+        }
+
+        /**
+         * Variance value.
+         */
+        public double variance() {
+            double mean = mean();
+            return (sumOfSquares / n) - mean * mean;
+        }
+
+        /**
+         * Standard deviation value.
+         */
+        public double std() {
+            return Math.sqrt(variance());
+        }
+
+        /**
+         * Min value.
+         */
+        public double min() {
+            return min;
+        }
+
+        /**
+         * Max value.
+         */
+        public double max() {
+            return max;
+        }
+    }
+
+    /**
+     * Class represents per feature statistics for impurity computing.
+     */
+    private class NodeStatistics {
+        /** Node id. */
+        private final NodeId nodeId;
+
+        /** Per feature statistics. */
+        private final Map<Integer, S> perFeatureStatistics;
+
+        /**
+         * Create an instance of NodeStatistics.
+         *
+         * @param nodeId Node id.
+         * @param perFeatureStatistics Per feature statistics.
+         */
+        public NodeStatistics(NodeId nodeId, Map<Integer, S> perFeatureStatistics) {
+            this.nodeId = nodeId;
+            this.perFeatureStatistics = perFeatureStatistics;
+        }
+
+        /**
+         * Store features statistics from other instance.
+         *
+         * @param other Other instance.
+         */
+        public void addOtherStatistics(Map<Integer, S> other) {
+            for (Integer featureId : other.keySet()) {
+                if (!perFeatureStatistics.containsKey(featureId))
+                    perFeatureStatistics.put(featureId, other.get(featureId));
+                else
+                    perFeatureStatistics.get(featureId).addHist(other.get(featureId));
+            }
+        }
+
+        /**
+         * Find best split point, based on feature statistics.
+         *
+         * @return Best split point if it exists.
+         */
+        public Optional<NodeSplit> findBestSplit() {
+            return perFeatureStatistics.values().stream()
+                .flatMap(x -> x.findBestSplit().map(Stream::of).orElse(Stream.empty()))
+                .min(Comparator.comparingDouble(NodeSplit::getImpurity));
+        }
+    }
+
+    /**
+     * Tree root class.
+     */
+    protected static class TreeRoot implements Model<Vector, Double> {
+        /** Root node. */
+        private TreeNode node;
+
+        /** Used features. */
+        private Set<Integer> usedFeatures;
```

Comment:
It's not clear where these `usedFeatures` are used?

Reply:
see code, please